### PR TITLE
feat: add support to keep leading and trailing slashes

### PR DESCRIPTION
**Pull request by ubmit**
This PR adds support to keep leading and trailing slashes if needed.

The idea here was to create a pipeline of "actions", if you will, that were being used to transform the `pathname`. My reasoning for this came from the need of triggering the `replaces` depending on `trailingSlash` and `leadingSlash` values.

Next steps: 
- unfortunately, I hit some corner cases and a few tests are still failing and I will analyze those soon;
- I still need to add a spec for the following case:
```js
const options = { trailingSlash: 'keep', leadingSlash: 'keep' }
```

### CODE REVIEW

File: src/index.js [hunk 1]
```diff
@@ -55,15 +68,15 @@ export default function urlJoin(...parts) {
     if (beforePathname) {
         url += beforePathname + (pathname ? '/' : '');
     // Otherwise start with the leading slash
-    } else if (options.leadingSlash) {
+    } else if (options.leadingSlash === true) {
```
Comment (satazor) [inline src/index.js]: This change is unrelated with the PR and it’s not really necessary.

File: src/index.js [hunk 2]
```diff
@@ -55,15 +68,15 @@ export default function urlJoin(...parts) {
...
         url += '/';
     }
 
     // Add pathname (foo/bar)
     url += pathname;
 
     // Add trailing slash
-    if (options.trailingSlash && !url.endsWith('/')) {
+    if (options.trailingSlash === true && !url.endsWith('/')) {
```
Comment (satazor) [inline src/index.js]: Same here

**Review by satazor — CHANGES_REQUESTED** [review]
The PR looks good! Let’s make sure all tests pass and we can merge it 💪

File: src/index.js [hunk 3]
```diff
@@ -3,19 +3,32 @@ import queryString from 'query-string';
...
-    .replace(/^\/+/, '')
-    // Remove trailing slashes
-    .replace(/\/+$/, '')
-    // Normalize consecutive slashes to just one
-    .replace(/\/+/g, '/');
+
+    const shouldKeep = (opt) => opt === 'keep';
+
+    const removeLeadingSlashes = (leadingSlash) =>
+        (s) => shouldKeep(leadingSlash) ? s : s.replace(/^\/+/, '');
```
Comment (satazor) [inline src/index.js]: I would name `s` as `pathname` instead to be more clear.

**Review by satazor — CHANGES_REQUESTED** [review]

**Comment by satazor** [comment]
@guilhermedeandrade The tests are failing because we are not doing the "beforePathname" + "pathname" concatenation correctly. As an example, `pathname` was guaranteed to not start with `/`, which is not the case anymore with `trailingSlash: 'keep'`.

My suggestion is to refactor the overall strategy a bit, so that `splitUrl` returns a more detailed data structure. Perhaps, the function should actually be renamed to `parseUrl`. Here's a few examples of how the output (data-structure) of that function would be:

- For `https://google.com/foo/bar?foz=baz`

```js
{
  prefix: 'https://google.com',
  pathname: {
    parts: ['foo', 'bar'],  // does not contain any / here
    hasLeading: true,
    hasTrailing: false,
  },
  suffix: '?foz=baz'
}
```

- For `//google.com/foo/bar?foz=baz` with `protocolRelative = true`:

```js
{
  prefix: '//google.com',
  pathname: {
    parts: ['foo', 'bar'],
    hasLeading: true,
    hasTrailing: false,
  },
  suffix: '?foz=baz'
}
```

- For `/foo/bar`

```js
{
  prefix: '',
  pathname: {
    parts: ['foo', 'bar'],
    hasLeading: true,
    hasTrailing: false,
  },
  suffix: ''
}
```

- For `foo/bar/`

```js
{
  prefix: '',
  pathname: {
    parts: ['foo', 'bar'],
    hasLeading: false,
    hasTrailing: true,
  },
  suffix: ''
}
```

Then, we would move the concatenation logic (joining the prefix + pathname parts + suffix) to the `urlJoin` function, or perhaps into a separate function. Because we now have all the information to recreate the URL based on the information returned from `splitUrl`, it's pretty straightforward to implement the concatenation the right way.

What do you think?

**Comment by ubmit** [comment]
> Instead of passing options to the actions, we could compose them as follows:
> 
> ```js
> const actions = [
>    leadingSlash !== 'keep' && removeLeadingSlashes,
>    trailingSlash !== 'keep' && removeTrailingSlashes,
>    normalizeConsecutiveSlashesToJustOne
> ].filter(Boolean)
> 
> const pathname = pipe(...actions)(match[2] || '')
> ```
> 
> This way, we remove complexity from actions' implementation and they become focused on their true purpose by not leaking options into them.
> 
> What do you think?

Good call, @acostalima ! It's an improvement for sure. Just pushed a commit where I refactored using your suggestion.

File: src/index.js (hunk 1 above)
Comment (ubmit) [reply src/index.js]: Oh, I don't know why I did it, sorry! Just fixed it

**Comment by satazor** [comment]
@guilhermedeandrade pinging in case you miss my comment above about why tests are failing.

**Comment by ubmit** [comment]
> @guilhermedeandrade pinging in case you miss my comment above about why tests are failing.

Haha ok, thanks! I was aware of your comment though, but I had to make a small pause! I'll get into it in a few minutes

File: src/index.js (hunk 3 above)
Comment (ubmit) [reply src/index.js]: This is not needed anymore, because the `removeLeadingSlashes` function is only taking the `leadingSlash` now

PS.: I made a small confusion, it was actually the other way around! The functions were only taking the `s` param! Anyway, just fixed it

File: src/index.js (hunk 2 above)
Comment (ubmit) [reply src/index.js]: Same reason as the comment above. It is fixed now!

**Comment by ubmit** [comment]
> @guilhermedeandrade The tests are failing because we are not doing the "beforePathname" + "pathname" concatenation correctly. As an example, `pathname` was guaranteed to not start with `/`, which is not the case anymore with `trailingSlash: 'keep'`.
> 
> My suggestion is to refactor the overall strategy a bit, so that `splitUrl` returns a more detailed data structure. Perhaps, the function should actually be renamed to `parseUrl`. Here's a few examples of how the output (data-structure) of that function would be:
> 
>     * For `https://google.com/foo/bar?foz=baz`
> 
> 
> ```js
> {
>   prefix: 'https://google.com',
>   pathname: {
>     parts: ['foo', 'bar'],  // does not contain any / here
>     hasLeading: true,
>     hasTrailing: false,
>   },
>   suffix: '?foz=baz'
> }
> ```
> 
>     * For `//google.com/foo/bar?foz=baz` with `protocolRelative = true`:
> 
> 
> ```js
> {
>   prefix: '//google.com',
>   pathname: {
>     parts: ['foo', 'bar'],
>     hasLeading: true,
>     hasTrailing: false,
>   },
>   suffix: '?foz=baz'
> }
> ```
> 
>     * For `/foo/bar`
> 
> 
> ```js
> {
>   prefix: '',
>   pathname: {
>     parts: ['foo', 'bar'],
>     hasLeading: true,
>     hasTrailing: false,
>   },
>   suffix: ''
> }
> ```
> 
>     * For `foo/bar/`
> 
> 
> ```js
> {
>   prefix: '',
>   pathname: {
>     parts: ['foo', 'bar'],
>     hasLeading: false,
>     hasTrailing: true,
>   },
>   suffix: ''
> }
> ```
> 
> Then, we would move the concatenation logic (joining the prefix + pathname parts + suffix) to the `urlJoin` function, or perhaps into a separate function. Because we now have all the information to recreate the URL based on the information returned from `splitUrl`, it's pretty straightforward to implement the concatenation the right way.
> 
> What do you think?

Seems like a good way to do it! I'll read the code a little bit more to remind me about how everything is behaving at the moment. If I have any suggestion to address this particular issue, I'll let you know; otherwise I'll just follow your advice accordingly @satazor

File: src/index.js [hunk 4]
```diff
@@ -3,22 +3,33 @@ import queryString from 'query-string';
...
-    const afterPathname = (match[3] || '');
-
-    return { beforePathname, pathname, afterPathname };
+    const prefix = match[1] || '';
+
+    const getParts = (prePathname) =>
+        prePathname.split('/').filter((part) => part !== '');
+
+    const hasLeadingSlash = (prePathname) =>
+        prePathname.match(/^\/+/) !== null;
```
Comment (satazor) [inline src/index.js]: You can use `regexp.test`, which is faster. See: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/RegExp/test

File: src/index.js [hunk 5]
```diff
@@ -3,22 +3,33 @@ import queryString from 'query-string';
...
+
+    const hasTrailingSlash = (prePathname) =>
+        prePathname.match(/\/+$/) !== null;
```
Comment (satazor) [inline src/index.js]: Same.

File: src/index.js [hunk 6]
```diff
@@ -45,30 +56,30 @@ export default function urlJoin(...parts) {
     .filter((part) => typeof part === 'string' || typeof part === 'number')
     .join('/');
 
-    // Split the parts into beforePathname, pathname, and afterPathname
+    // Split the parts into prefix, pathname, and suffix
     // (scheme://host)(/pathname)(?queryString)
-    const { beforePathname, pathname, afterPathname } = splitUrl(partsStr, options);
+    const { prefix, pathname, suffix } = parseUrl(partsStr, options);
```
Comment (satazor) [inline src/index.js]: The code below needs now to take into consideration the `pathname.hasLeadingSlash`, `pathname.hasTrailingSlash`, `options.leadingSlash` and `options.trailingSlash` to correctly build the final url.

Comment (ubmit) [reply src/index.js]: I've just added more code to deal with the options `leadingSlash` and `trailingSlash`, but I'd like to check one thing:

For `/foo/bar/` with `options.trailingSlash = 'keep'` and `options.leadingSlash = 'keep'`, the correct output of the `parseUrl` function would be this one below?
```js
 {
   prefix: '',
   pathname: {
     parts: [ '/foo', 'bar/' ],
     hasLeading: true,
     hasTrailing: true
  },
  suffix: ''
}
```

Comment (satazor) [reply src/index.js]: `parseUrl` should always contain no slashes in the parts:

 ```js
 {
   prefix: '',
   pathname: {
     parts: [ 'foo', 'bar' ],
     hasLeading: true,
     hasTrailing: true
  },
  suffix: ''
}
```

The concatenation should be done in the `urlJoin` function or perhaps into a new `recreateUrl` / `rebuildUrl` function. In the case you mentioned, because it knows that it should preserve both trailing and leading, it would respect what `pathname.hasLeading` and `pathname.hasTrailing` says. In this case, it would recreate the following URL: `/foo/bar/`.

Comment (satazor) [reply src/index.js]: But, as an example, if `options.leadingSlash` was true and `options.trailingSlash` was false, it would recreate the following URL: `/foo/bar` (it would ignore what `pathname.hasTrailing` says).

Comment (ubmit) [reply src/index.js]: Oh ok! My bad, I've just commited the wrong way of doing it. I'll undo this last commit and start working on this `recreateUrl` / `rebuildUrl`. Thanks for the info

Comment (satazor) [reply src/index.js]: No problem! If we were chatting personally, these misunderstandings wouldn't happen 👍

File: src/index.js (hunk 4 above)
Comment (ubmit) [reply src/index.js]: Fixed

File: src/index.js (hunk 5 above)
Comment (ubmit) [reply src/index.js]: Fixed

File: src/index.js [hunk 7]
```diff
@@ -55,24 +55,28 @@ export default function urlJoin(...parts) {
...
+    const { leadingSlash, trailingSlash } = options;
+
     let url = '';
 
     // Start with prefix if not empty (http://google.com)
     if (prefix) {
-        url += prefix + (pathname.parts.length > 0 ? '/' : '');
-    // Otherwise start with the leading slash
-    } else if (options.leadingSlash) {
+        url += prefix + (pathnameParts.length > 0 ? '/' : '');
```
Comment (ubmit) [inline src/index.js]: I'm really not sure if I should change this line or not, is this the expected behavior?

File: src/index.js (hunk 6 above)
Comment (ubmit) [reply src/index.js]: Could you please confirm me which are the expected outputs for the following cases?

1. When `options = { trailingSlash: true }`, what `urlJoin(options)` should return?
2. When `options = { trailingSlash:  'keep' }`, what `urlJoin(options)` should return?
3. When `options = { leadingSlash:  'keep' }`, what `urlJoin(options)` should return?

Comment (satazor) [reply src/index.js]: 1. `/`
2. `<empty>`
3. `<empty>`

Comment (ubmit) [reply src/index.js]: >     1. `/`
> 
>     2. `<empty>`
> 
>     3. `<empty>`

Got it! Thanks

File: src/index.js (hunk 7 above)
Comment (satazor) [reply src/index.js]: I think it's correct.

File: test/index.spec.js [hunk 8]
```diff
@@ -225,3 +225,73 @@ it('should handle the provided query and query options objects', () => {
...
+    expect(urlJoin('http://google.com', 'foo', options)).toBe('http://google.com/foo');
+    expect(urlJoin('http://google.com/', 'foo', options)).toBe('http://google.com/foo');
+    expect(urlJoin('http://google.com/', '/foo', options)).toBe('http://google.com/foo');
+    expect(urlJoin('http://google.com//', '/foo', options)).toBe('http://google.com/foo');
+    expect(urlJoin('http://google.com/foo', 'bar', options)).toBe('http://google.com/foo/bar');
+
+    expect(urlJoin('http://google.com', '?queryString', options)).toBe('http://google.com?queryString');
+    expect(urlJoin('http://google.com', 'foo?queryString', options)).toBe('http://google.com/foo?queryString');
+    expect(urlJoin('http://google.com', 'foo', '?queryString', options)).toBe('http://google.com/foo?queryString');
+    expect(urlJoin('http://google.com', 'foo/', '?queryString', options)).toBe('http://google.com/foo/?queryString');
```
Comment (satazor) [inline test/index.spec.js:376]: This test is invalid, should be: 

```js
expect(urlJoin('http://google.com', 'foo/', '?queryString', options)).toBe('http://google.com/foo?queryString');
```

**Comment by satazor** [comment]
@guilhermedeandrade This issue was indeed more complex than estimated. I took the liberty to help you out. This gist has all tests passing: https://gist.github.com/satazor/c473756fadbb54826ce6004f297c7121

Note that you have an error in the tests, I've already commented out, so you should fix it first.

Feel free to modify the Gist to make it more functional.

File: src/index.js (hunk 6 above)
Comment (ubmit) [reply src/index.js]: > Could you please confirm me which are the expected outputs for the following cases?
> 
>     1. When `options = { trailingSlash: true }`, what `urlJoin(options)` should return?
> 
>     2. When `options = { trailingSlash:  'keep' }`, what `urlJoin(options)` should return?
> 
>     3. When `options = { leadingSlash:  'keep' }`, what `urlJoin(options)` should return?

I was able to address the first case, but I'm stuck on the cases 2 and 3.

Case 2: the problem here is that at line 54 the leading slash is being added and it simply remains there, so the final output will be `/` instead of `<empty>`. I thought about adding another condition on the `addLeadingSlash` function, so it would leave this context without the slash, but it would still be added by the `addTrailingSlash` later on the pipeline. I'm not sure which is the better approach here, where I should step in.

Case 3: the problem here is in the way I created the `addTrailingSlash` function it will add the slash if the current url does not end with `/`.

Any suggestion on how should I fix those?

Comment (satazor) [reply src/index.js]: See my comment below on the PR

Comment (ubmit) [reply src/index.js]: Oh ok, nevermind then! Thanks!

File: test/index.spec.js (hunk 8 above)
Comment (ubmit) [reply test/index.spec.js:376]: Fixed!

File: test/index.spec.js [hunk 9]
```diff
@@ -225,3 +225,73 @@ it('should handle the provided query and query options objects', () => {
 
     expect(urlJoin('/google.com', options)).toBe('/google.com?foo[0]=1&foo[1]=2&foo[2]=3');
 });
+
+it('should keep the trailing slash', () => {
+    const options = { trailingSlash: 'keep' };
+
+    expect(urlJoin(options)).toBe('');
```
Comment (ubmit) [inline test/index.spec.js:236]: This is still failing though @satazor

Comment (satazor) [reply test/index.spec.js:236]: Sorry, that test is wrong as well. Because, `leadingSlash` is true by default, it should be `/`.

See test on line 4.

Comment (satazor) [reply test/index.spec.js:236]: @guilhermedeandrade I've updated the gist, fixing a few bugs in the code and also added a test file based on your tests but with more combinations.

Comment (ubmit) [reply test/index.spec.js:236]: Ok, fixed!

Comment (ubmit) [reply test/index.spec.js:236]: @satazor Oh sorry, when I was typing my last comment your latest one had not appeared yet. I will update and push the changes

Comment (satazor) [reply test/index.spec.js:236]: Maybe I didn't updated at all. I've pushed the gist now.

Comment (ubmit) [reply test/index.spec.js:236]: Just pushed your changes @satazor

**Review by satazor — APPROVED** [review]
